jit-compiler/src/compiler.rs

```diff
@@ -104,28 +110,35 @@ pub fn call_cargo(code: &str) -> Result<(Temp, String), String> {
         .join("target")
         .join("release")
         .join(format!("libpowdr_jit_compiled.{extension}"));
-    Ok((dir, lib_path.to_str().unwrap().to_string()))
+    Ok(PathInTempDir {
+        dir,
+        path: lib_path.to_str().unwrap().to_string(),
+    })
 }
 
-/// Loads the given library and creates funtion pointers for the given symbols.
-pub fn load_library(path: &str, symbols: &[&str]) -> Result<SymbolMap, String> {
-    let c_path = CString::new(path).unwrap();
-    let lib = unsafe { dlopen(c_path.as_ptr(), RTLD_NOW) };
-    if lib.is_null() {
-        return Err(format!("Failed to load library: {path:?}"));
-    }
-    let mut result = HashMap::new();
-    for sym in symbols {
-        let extern_sym = extern_symbol_name(sym);
-        let sym_cstr = CString::new(extern_sym).unwrap();
-        let fun_ptr = unsafe { dlsym(lib, sym_cstr.as_ptr()) };
-        if fun_ptr.is_null() {
-            return Err(format!("Failed to load symbol: {fun_ptr:?}"));
-        }
-        let fun = unsafe { std::mem::transmute::<*mut c_void, fn(u64) -> u64>(fun_ptr) };
-        result.insert(sym.to_string(), fun);
-    }
-    Ok(result)
+/// Loads the given library and creates function pointers for the given symbols.
+pub fn load_library(
+    path: &str,
+    symbols: &[&str],
+) -> Result<HashMap<String, LoadedFunction>, String> {
+    let library = Arc::new(
+        unsafe { libloading::Library::new(path) }
+            .map_err(|e| format!("Error loading library at {path}: {e}"))?,
+    );
+    symbols
+        .iter()
+        .map(|&sym| {
+            let extern_sym = extern_symbol_name(sym);
+            let function =
+                *unsafe { library.get::<unsafe extern "C" fn(u64) -> u64>(extern_sym.as_bytes()) }
+                    .map_err(|e| format!("Error accessing symbol {sym}: {e}"))?;
+            let fun = LoadedFunction {
+                library: library.clone(),
+                function,
+            };
+            Ok((sym.to_string(), fun))
+        })
+        .collect::<Result<_, String>>()
```

Comment:
Maybe this entire block should go inside an `SymbolMap::get()`.

Reply:
You mean we should delay accessing the symbol? I think it's better to find them right away first. The issue is that for fixed column creation, I would like to have a proper function pointer that has the least cost associated to it because we call it 2**20 times, and I don't want to search for it 2**20 times.

Reply:
I don't think this would be any different. In the current way, at some point you will have to retrieve the `LoadedFunction` from the `HashMap` returned by `load_library()`. In the way I propose, you will instead retrieve the `libloading::Symbol` from the `SymbolMap` returned by `load_library()` and still will have a proper function pointer to call, just that it is lifetime bound to the `SymbolMap` instead of being freestanding (due to the lack of Arc).

Reply:
Depending how it is used, the lifetime can be an inconvenience, sure, in this case your solution is better.

Reply:
At some point, we will store the type of the symbol. I think at latest then we can't really expose the Symbol stuff to the user either.

Another point I would like to make: If we return a Symbol to the user, there is no protection against dereferencing the symbol. Once the symbol is dereferenced into a function, it can be freely copied, and the association with the library is gone. That's why I think not exposing the function type to the user at all is the best way.